capsules/src/proximity.rs

```diff
@@ -0,0 +1,310 @@
+                }
+
+                app.subscribed = true; // enqueue
+                app.enqueued_command_type = command;
+
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into());
+
+        if r == ReturnCode::EBUSY {
+            return ReturnCode::EBUSY;
+        }
+
+        // Only run command if it is only one in queue otherwise we wait for callback() for last run command to trigger another command to run
+        let mut num_commands: u8 = 0;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if app.subscribed {
+                    num_commands += 1;
+                }
+            });
+        }
+        if num_commands == 1 {
+            self.run_next_command();
+        }
+
+        ReturnCode::SUCCESS
+    }
+
+    fn run_next_command(&self) -> ReturnCode {
+        let mut break_flag: bool = false;
+
+        // Find and run another command
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if app.subscribed {
+                    // run it
+                    match app.enqueued_command_type {
+                        ProximityCommand::ReadProximity => {
+                            self.call_driver(app.enqueued_command_type, 0, 0);
+                        }
+                        ProximityCommand::ReadProximityOnInterrupt => {
+                            let t: Thresholds = self.find_thresholds();
+                            self.call_driver(
+                                app.enqueued_command_type,
+                                t.lower as usize,
+                                t.upper as usize,
+                            );
+                        }
+                        _ => {}
+                    }
+
+                    break_flag = true;
+                }
+            });
+
+            if break_flag {
+                break;
+            }
+        }
+
+        ReturnCode::SUCCESS
+    }
+
+    fn find_thresholds(&self) -> Thresholds {
+        // Get the lowest upper prox and highest lower prox of all subscribed apps
+        // With the IC thresholds set to these two values, we ensure to never miss an interrupt-causing proximity value for any of the
+        // apps
+        let mut highest_lower_proximity: u8 = 0;
+        let mut lowest_upper_proximity: u8 = 255;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if (app.lower_proximity > highest_lower_proximity) && app.subscribed {
+                    highest_lower_proximity = app.lower_proximity;
+                }
+                if (app.upper_proximity < lowest_upper_proximity) && app.subscribed {
+                    lowest_upper_proximity = app.upper_proximity;
+                }
+            });
+        }
+
+        // return values
+        Thresholds {
+            lower: highest_lower_proximity,
+            upper: lowest_upper_proximity,
+        }
```

Comment:
Shouldn't this set the thresholds of the underlying driver to be the lowest lower proximity and the highest upper proximity value? 

Ex:
app1: lower_prox=10, higher_prox = 20
app2: lower_prox=5, higher_prox = 25

It seems the current algorithm would set thresholds of 10/20 in this case, thereby missing readings between 5 and 10 and 20/25 that app2 would actually want to detect.

Perhaps I am misunderstanding?

Reply:
the sensor is set up in such a way that an interrupt will fire if (reading < 10) or (reading > 20).  If reading == 8 for example, an interrupt will fire and the callback code will alert app1 as it requires this information.  However, the callback code would not alert app2 because its thresholds indicate that it does not require this information.  Lines 288-289 in proximity.rs outline the "callback code" I am referring to.

Reply:
Oh, I see! My mistake.